### PR TITLE
Identity pallet improvements

### PR DESCRIPTION
**Pull request by georgepisaltu**
This PR is a follow up to #1661 

- [x] rename the `simple` module to `legacy`
- [x] fix benchmarks to disregard the number of additional fields
- [x] change the storage deposits to charge per encoded byte of the identity information instance, removing the need for `fn additional(&self) -> usize` in `IdentityInformationProvider`
- [x] ~add an extrinsic to rejig deposits to account for the change above~
- [ ] ~ensure through proper configuration that the new byte-based deposit is always lower than whatever is reserved now~
- [x] remove `IdentityFields` from the `set_fields` extrinsic signature, as per [this discussion](https://github.com/paritytech/polkadot-sdk/pull/1661#discussion_r1371703403)

> ensure through proper configuration that the new byte-based deposit is always lower than whatever is reserved now

Not sure this is needed anymore. If the new deposits are higher than what is currently on chain and users don't have enough funds to reserve what is needed, the extrinisc fails and they're basically grandfathered and frozen until they add more funds and/or make a change to their identity. This behavior seems fine to me. Original idea [here](https://github.com/paritytech/polkadot-sdk/pull/1661#issuecomment-1779606319).

> add an extrinsic to rejig deposits to account for the change above

This was initially implemented but now removed from this PR in favor of the implementation detailed [here](https://github.com/paritytech/polkadot-sdk/pull/2088).

### CODE REVIEW

File: substrate/frame/identity/src/lib.rs
Comment (joepetrowski) [inline substrate/frame/identity/src/lib.rs:104]: This seems very large, I imagine even 1 kB would be enough. Maybe 5 kB?

File: substrate/frame/identity/src/lib.rs
Comment (joepetrowski) [inline substrate/frame/identity/src/lib.rs:345]:
```suggestion
			let byte_deposit = T::ByteDeposit::get().saturating_mul(<BalanceOf<T>>::from(encoded_byte_size));
```

File: substrate/frame/identity/src/lib.rs
Comment (joepetrowski) [inline substrate/frame/identity/src/lib.rs:362]:
```suggestion
			id.deposit = T::BasicDeposit::get().saturating_add(byte_deposit);
```

File: substrate/frame/identity/src/lib.rs
Comment (joepetrowski) [inline substrate/frame/identity/src/lib.rs:910]: I mean either one is fine, but this will get in before #1814 
```suggestion
		#[pallet::call_index(15)]
```

File: substrate/frame/identity/src/lib.rs
Comment (joepetrowski) [inline substrate/frame/identity/src/lib.rs:927]:
```suggestion
						T::ByteDeposit::get().saturating_mul(<BalanceOf<T>>::from(encoded_byte_size));
```

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:104]: I just took the constant from the preimage pallet. Here is a more informed view: a "full" `IdentityInfo` without any entries in `additional` has an encoded size of 253 bytes. Each entry in `additional` is then worth at most 66 bytes, depending on the values chosen for the key and value, since they're `Data`. With the current values for `T::MaxAdditionalFields` of 100, that brings us to at most `253 + 66 * 100` or ~6.7 KiB maximum encoded size for any value currently in storage. Considering this, I set it to 7 KiB now and we can lower it later if we wish to do so.

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:345]: Done.

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:362]: Done.

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:910]: Done.

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:927]: Done.

**Review by georgepisaltu — COMMENTED** [review]
I also added some more checked arithmetic where applicable, just to be consistent.

File: substrate/frame/identity/src/lib.rs
Comment (joepetrowski) [reply substrate/frame/identity/src/lib.rs]: Yeah, that sounds good. The preimage pallet is meant to store data such that you can remove it if the system hasn't requested it. Setting an identity clears any judgements, so we don't really expect people to remove their data once they have set it. Of course the byte fee accounts for it, but it seems more flexible to just have a field meant to take a hash that can be hosted elsewhere. Then, people could actually use the preimage pallet if it needed to be on-chain, or some off-chain hosting service if the reader of the Identity Info knew where to look for it.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4111258

**Review by georgepisaltu — COMMENTED** [review]
> remove IdentityFields from the `set_fields` extrinsic signature, as per https://github.com/paritytech/polkadot-sdk/pull/1661#discussion_r1371703403

The `set_fields` extrinisc now accepts the underlying numerical representation as the input. This is currently `u64` for us and will probably continue to be since there's no reason to change it. This change requires no migration now, since it's the same `u64` in storage as before, but if the underlying numeric type changes when we move away from the `legacy` impls, it will require a migration.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4143061

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4143092

File: substrate/bin/node/runtime/src/lib.rs
Comment (joepetrowski) [inline substrate/bin/node/runtime/src/lib.rs:1464]: For Rococo and Westend we can leave them since we'll drop the pallet from the Relay Chains and configure it correctly in genesis of the parachain, but here (being the example runtime), I think we should use `deposit()` for the `BasicDeposit` and `SubAccountDeposit` configs using the smallest encoded size of a set identity.

File: substrate/frame/identity/src/lib.rs
Comment (joepetrowski) [inline substrate/frame/identity/src/lib.rs:246]:
```suggestion
		/// Identity encoded size is too large.
```

**Review by joepetrowski — APPROVED** [review]
Should also get review from @bkchr and/or @ggwpez before merging though

File: substrate/bin/node/runtime/src/lib.rs
Comment (georgepisaltu) [reply substrate/bin/node/runtime/src/lib.rs:1464]: To answer the obvious questions first: the smallest identity you can set is an empty one, encoding to 9 bytes. The whole `Registration` `struct` that actually contains the identity would encode to a bit more, but it has stuff we don't charge for right now, such as the deposit held and the judgements.

That being said, I think of this deposit a bit differently. First off, even with the smallest identity, those 9 bytes will be accounted for because we take the encoded size of the whole `struct`, not just the data within it. Right now, the base deposit is way bigger than what it actually needs to be to store the data on chain, and I think that's an intentional barrier of entry into the pallet right now. I don't know if the deposits *need* to be this big because of some other economic reasons. My point is this: the storage cost should already be accounted for by the byte deposit, so the base deposit shouldn't depend on the `struct` size. It should instead be tied to some economic reason and if we can't find one, we can remove it or set it to `0` on our implementations and let downstream users decide if they want to add such a cost.

Comment (joepetrowski) [reply substrate/bin/node/runtime/src/lib.rs:1464]: I should have checked before commenting, but on Polkadot it's already configured [like I suggested](https://github.com/polkadot-fellows/runtimes/blob/7bd91e826e4b2d95f6d2ee7f6bb0f0f47d095859/relay/polkadot/src/lib.rs#L665). On Kusama, [not](https://github.com/polkadot-fellows/runtimes/blob/7bd91e826e4b2d95f6d2ee7f6bb0f0f47d095859/relay/kusama/src/lib.rs#L919).

By using `deposit`, it does actually come out to over 20 DOT:
```python
>>> def deposit(items, bytes):
...     return items * 20 * 10_000_000_000 + bytes * 100 * 100_000
... 
>>> deposit(1, 9) / 10**10
20.009
```

You aren't just paying for the size in bytes, you're paying for having a storage item at all, because more items increases the depth of the trie and therefore increases weights for all transactions since reads, writes, and trie hashing take longer.

For sure other users can configure this with a higher or lower `BasicDeposit` to subsidize or discourage its use, but I think the neutral config is to set it to `(1, min_size(Registration))`.

Comment (georgepisaltu) [reply substrate/bin/node/runtime/src/lib.rs]: The `BasicDeposit` for identity is now `deposit(1, min_size(Registration) - min_size(IdentityInfo)`, because the bytes in the identity info are already accounted for in the byte deposit. I didn't replicate the change for the subs deposit because if we were to do it, that would be another 2 storage items: one entry in `SuperOf` per sub identity of an account, one entry in `SubsOf` per account which uses the sub-account functionality. On top of that, we would have to also charge per byte and the price for a sub-identity would end up being about twice the price of setting one identity. Right now, the price of a sub-identity is 1/5 of an identity. I'd prefer to keep `SubAccountDeposit` as is, let me know what you think.

File: substrate/frame/identity/src/legacy.rs
Comment (ggwpez) [inline substrate/frame/identity/src/legacy.rs:28]: Why can we not use the correct maximum from the runtime?

File: substrate/frame/identity/src/lib.rs
Comment (ggwpez) [inline substrate/frame/identity/src/lib.rs:106]: Where does this come from and why is it private?  
is this statically enforced by testing that `Identity::max_encoded_len() <= IDENTITY_MAX_SIZE`?

Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:106]: Origin [here](https://github.com/paritytech/polkadot-sdk/pull/2048#discussion_r1373512069). In short, it's the same model as in [preimage](https://github.com/paritytech/polkadot-sdk/blob/9987bbb1d264672d406071fb2e164e4e80117e13/substrate/frame/preimage/src/lib.rs#L94) but adapted to what is currently stored in this pallet.

> why is it private?

I guess it could be public. My reasoning was that users choose what `struct` to use for their `IdentityInformationProvider` and this value of the max size is unreasonably large anyway. If it didn't fail the decoding, then we'll have a hard stop in the pallet regarding the size.

> is this statically enforced by testing that `Identity::max_encoded_len() <= IDENTITY_MAX_SIZE`?

If you mean in integrity test, then not yet but it's a very good idea.

File: substrate/frame/identity/src/legacy.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/legacy.rs:28]: We surely can. I was getting rid of the `MaxAdditionalFields` instances from the configs and I forgot to get rid of this, good catch!

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs]: Done, PTAL

File: substrate/frame/identity/src/types.rs
Comment (bkchr) [inline substrate/frame/identity/src/types.rs:247]:
```suggestion
	/// Check if an identity registered information for some given `fields`.
	fn has_identity(&self, fields: Self::IdentityField) -> bool;
```

`IdentityField` should get a better name.

File: substrate/frame/identity/src/types.rs
Comment (bkchr) [inline substrate/frame/identity/src/types.rs:40]:
```suggestion
```

We don't need this.

File: substrate/frame/identity/src/lib.rs
Comment (bkchr) [inline substrate/frame/identity/src/lib.rs:674]:
```suggestion
			fields: <T::IdentityInformation as IdentityInformationProvider>::IdentityField,
```

File: substrate/frame/identity/src/lib.rs
Comment (bkchr) [inline substrate/frame/identity/src/lib.rs:687]:
```suggestion
					registrar.fields = fields;
```

**Review by bkchr — COMMENTED** [review]
Ty! Looking good, but still requires some improvements.

The identity fields stuff is still too much "married" to the old implementation. You are doing there some hops to let it look like some bitfield, while this isn't required. The `IndentityProvider` should just declare what kind of type this is and then it can also do internally the `has_identity` call. However, this will also require that this code: https://github.com/paritytech/polkadot-sdk/blob/9ca267328e0cba753aa528615cc19d1f633ed764/substrate/frame/alliance/src/lib.rs#L1084-L1107

is made generic. TBH, it doesn't make sense that `alliance` by default declares what kind of fields are required. This kind of configuration should be done in the runtime.

**Comment by bkchr** [comment]
I know that this alliance stuff was there already before. However, we should do it correctly now.

File: substrate/frame/identity/src/lib.rs
Comment (joepetrowski) [reply substrate/frame/identity/src/lib.rs:374]: You still have to account for the storage item existing. This pattern is common:

- [Preimage](https://github.com/polkadot-fellows/runtimes/blob/7bd91e8/relay/polkadot/src/lib.rs#L229) (pre-`Consideration`)
- [Multisig](https://github.com/polkadot-fellows/runtimes/blob/7bd91e8/relay/polkadot/src/lib.rs#L917)
- [Proxy](https://github.com/polkadot-fellows/runtimes/blob/7bd91e8/relay/polkadot/src/lib.rs#L935)

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:106]: Moved it to config.

File: substrate/frame/identity/src/types.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/types.rs:40]: Removed it.

File: substrate/frame/identity/src/types.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/types.rs:247]: @joepetrowski suggested `FieldsBitFlags` and I don't have a better idea :sweat_smile:

File: substrate/frame/alliance/src/lib.rs
Comment (georgepisaltu) [inline substrate/frame/alliance/src/lib.rs:144]: These 2 functions could be collapsed into `has_required_identities(who: &AccountId, fields: Self::FieldsBitFlags) -> bool` but I wasn't sure it's better than explicitly declaring the relevant fields in a standalone function as it is now. Let me know what you think.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4194481

File: substrate/frame/identity/src/types.rs
Comment (georgepisaltu) [inline substrate/frame/identity/src/types.rs:235]: `BitFlags` had the advantage of validating the input automatically and now we lost that. I was thinking we could add a function in this trait `fn is_valid(fields: Self::FieldsBitFlags) -> bool` to validate input so we don't end up with people setting fields that don't exist.

**Comment by georgepisaltu** [comment]
> I know that this alliance stuff was there already before. However, we should do it correctly now.

In addition to the changes in `pallet-identity` to decouple from `BitFlags`, I also made the requested changes in `pallet-alliance`. @bkchr PTAL

File: substrate/frame/alliance/src/lib.rs
Comment (bkchr) [reply substrate/frame/alliance/src/lib.rs:144]: Ha nice! I had the same idea :D And I think we should do it this way. No need to have these two functions!

File: substrate/frame/alliance/src/lib.rs
Comment (bkchr) [inline substrate/frame/alliance/src/lib.rs:137]: Then we can also get rid off this.

File: substrate/frame/identity/src/lib.rs
Comment (bkchr) [inline substrate/frame/identity/src/lib.rs:262]: Not being used?

File: substrate/frame/identity/src/types.rs
Comment (bkchr) [inline substrate/frame/identity/src/types.rs:243]:
```suggestion
	/// Type capable of holding information on which fields are set.
	type FieldsInfo: sp_runtime::traits::Member + Default;
```

Comment (bkchr) [reply substrate/frame/identity/src/types.rs:243]: Naming is hard and I still don't like my name. Maybe @joepetrowski has a better idea.

File: substrate/frame/identity/src/lib.rs
Comment (bkchr) [inline substrate/frame/identity/src/lib.rs:295]: This should always fail, because the `max_encoded_len` of the `IdentityInforrmation` would be with all fields set at their maximum.

File: substrate/frame/identity/src/lib.rs
Comment (bkchr) [inline substrate/frame/identity/src/lib.rs:139]: Not sure we need this at all? Why have an upper bound here, if this upper bound is already implicitly declared by the `IdentityInformation`.

I don't think we need this.

Comment (bkchr) [reply substrate/frame/identity/src/lib.rs:139]: A chain that has 10 different identity fields, should also support that all of them are set.

**Review by bkchr — APPROVED** [review]
Thank you! Generally looking really good now! Left some last nitpicks, but overall good work!

File: substrate/frame/identity/src/lib.rs
Comment (ggwpez) [reply substrate/frame/identity/src/lib.rs:375]: We should probably be using storage `Consideration`, which abstracts the `num_bytes -> Balance` conversion.

File: substrate/frame/identity/src/types.rs
Comment (bkchr) [reply substrate/frame/identity/src/types.rs:243]: Ahh just have seen: https://github.com/paritytech/polkadot-sdk/pull/2048#discussion_r1380637556

However, we should not include "BitFlags" in the name, because it could be everything (also may a vector for example).

Comment (joepetrowski) [reply substrate/frame/identity/src/types.rs:243]: > However, we should not include "BitFlags" in the name, because it could be everything (also may a vector for example).

Yeah good point. `FieldsIdentifier`? This one is difficult to name.

File: substrate/frame/alliance/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/alliance/src/lib.rs:137]: Done.

File: substrate/frame/alliance/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/alliance/src/lib.rs:144]: Done.

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:103]: Done.

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:139]: Makes sense. It's gone.

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:262]: Removed.

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:295]: Gone now with the other change.

File: substrate/frame/identity/src/types.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/types.rs:243]: Renamed it to `FieldsIdentifier`. Maybe we'll come up with a better name in the future :grin:

File: substrate/frame/identity/src/types.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/types.rs:252]: Done.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4205213

**Pull request merged**
